# Problem 2

In [1]:
#Import the packages
import pandas as mypd
from sklearn import tree
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import GridSearchCV
import matplotlib.pyplot as myplot
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
import math as mymath
import seaborn as mysb
from sklearn.ensemble import RandomForestClassifier
import random
from statsmodels.formula.api import ols
from scipy import stats
import statsmodels.api as mysm
from sklearn.linear_model import LogisticRegression
import numpy as mynp
from sklearn.metrics import confusion_matrix

In [2]:
#Import the dataset
mydata = mypd.read_excel('/Users/sudiptarakshit/Desktop/ISI MS QMS/Sem2/MDA/Assignment/Assignment6/Heart_Disease_Data.xlsx')

In [3]:
#Explore the dataset
mydata.head()

Age  Sex  CP  RestBP  Cholesteral  FBP  RestECG  Max_HR  ExAngina  Oldpeak  \
0   63    1   3     145          233    1        0     150         0      2.3   
1   37    1   2     130          250    0        1     187         0      3.5   
2   41    0   1     130          204    0        0     172         0      1.4   
3   56    1   1     120          236    0        1     178         0      0.8   
4   57    0   0     120          354    0        1     163         1      0.6   

   Slope  CA  Thal  Result  
0      0   0     1       1  
1      0   0     2       1  
2      2   0     2       1  
3      2   0     2       1  
4      2   0     2       1

In [4]:
#Check the dataset
mydata.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Age          303 non-null    int64  
 1   Sex          303 non-null    int64  
 2   CP           303 non-null    int64  
 3   RestBP       303 non-null    int64  
 4   Cholesteral  303 non-null    int64  
 5   FBP          303 non-null    int64  
 6   RestECG      303 non-null    int64  
 7   Max_HR       303 non-null    int64  
 8   ExAngina     303 non-null    int64  
 9   Oldpeak      303 non-null    float64
 10  Slope        303 non-null    int64  
 11  CA           303 non-null    int64  
 12  Thal         303 non-null    int64  
 13  Result       303 non-null    int64  
dtypes: float64(1), int64(13)
memory usage: 33.3 KB


We find that there are total 303 entries in the dataset and there is no missing value present.

There are 13 expalnatory features in the data. 

All the features are numeric type. Some of the features are ordinal type discrete numeric data apparently.

In [5]:
#Separate x's and y
x = mydata.iloc[:,0:13]
y = mydata.iloc[:,13]

## Split the data randomly into training (80%) and test (20%)

We initialize the random number generator.

In [6]:
#Set seed
random.seed(100)

In [7]:
#Split the data into training (80%) and testing (20%)
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size=0.2,random_state=100)

In [8]:
#Check the test/train split
x_train.count(), x_test.count()

(Age            242
 Sex            242
 CP             242
 RestBP         242
 Cholesteral    242
 FBP            242
 RestECG        242
 Max_HR         242
 ExAngina       242
 Oldpeak        242
 Slope          242
 CA             242
 Thal           242
 dtype: int64,
 Age            61
 Sex            61
 CP             61
 RestBP         61
 Cholesteral    61
 FBP            61
 RestECG        61
 Max_HR         61
 ExAngina       61
 Oldpeak        61
 Slope          61
 CA             61
 Thal           61
 dtype: int64)

train data has 242 observations & test data has 61 observations.

| | **Operation on Train data** |  |
| :----: | :----: | :----: |
| | Build a General Model |  |
| |⇩| |
| |R<sup>2</sup> Accuracy Check | |
| |⇩| |
| |Cross Validation R<sup>2</sup> Check for generalizability | |
| |⇩ | |
| |Hyper Parameter Tuning| |
| |⇩ | |
| |Build model using the best parameters| |
| |⇩ | |
| |R<sup>2</sup> Accuracy Check | |
| |⇩ | |
| |Cross Validation R<sup>2</sup> Check for generalizability | |
| |⇩ | |
||Get predicted value on train data||
| |⇩ | |
|| Confusion Matrix||
| |⇩ | |
||Accuracy on Training Data||
| |⇩ | |
||Missclassification on training data||

| | **Operation on Test data** |  |
| :----: | :----: | :----: |
||Get predicted value on Test data||
| |⇩| |
|| Confusion Matrix||
| |⇩ | |
||Accuracy on Test Data||
| |⇩ | |
||Missclassification on Test data||

## Classification Tree Method (CART)

In [9]:
#Specify the model
mymodel = tree.DecisionTreeClassifier(min_samples_split=10)

In [10]:
#Fit the model to training data
mymodel = mymodel.fit(x_train,y_train)

In [11]:
#Model accuarcy
accuracy = mymodel.score(x_train,y_train)
round(accuracy,3)

0.921

There may be an overfitting. Model generalizability can be less. We check model generalizability using cross validation.

**K fold cross validation**

A model is trained using k-1 of the folds as training data

The resulting model is validated on the remaining part of the data (i.e., it is used as a test set to compute a performance measure such as accuracy)

Then we take the average accuracy.

In [12]:
#Perform a 4 fold Cross Validation
myscore = cross_val_score(estimator=mymodel,X=x_train,y=y_train,cv=4,scoring='accuracy')

In [13]:
#Cross Validation Accuracy
myscore.mean()

0.7562158469945355

We see a significant drop in model accuracy after cross validation. We have to do hyper-parameter tuning for a more generalized model.

**Hyper parameter optimization**

In [14]:
#Hyper parameter tuning
myparamgrid = [{'min_samples_split':[5,10,15,20,25,30,35],'criterion':['gini','entropy']}]

In [15]:
# Set up search
mysearch = GridSearchCV(estimator=mymodel,param_grid=myparamgrid,scoring='accuracy',cv=5,verbose=True)

In [16]:
#Fit the search on training data
mysearch = mysearch.fit(X=x_train,y=y_train)

Fitting 5 folds for each of 14 candidates, totalling 70 fits


In [17]:
#Optimum parameter values
mysearch.best_params_

{'criterion': 'entropy', 'min_samples_split': 10}

In [18]:
#Specify optimum model
mymodel = tree.DecisionTreeClassifier(criterion='entropy',min_samples_split=10)

In [19]:
# Fit the model to training data
mymodel = mymodel.fit(X=x_train,y=y_train)

In [20]:
# Check the model accuracy
accuracy = mymodel.score(x_train,y_train)
accuracy

0.9338842975206612

There is no improvement in model accuracy after hyper parameter optimization, Also, there should be increase in model generalizability.

In [22]:
#Cross validation accuracy
myscore = cross_val_score(estimator=mymodel,X=x_train,y=y_train,scoring='accuracy',cv=5)
myscore.mean()

0.7440476190476191

We see that cross validation accuracy has also increased.

In [23]:
# Predicted Values on train data
predy = mymodel.predict(X=x_train)

In [24]:
#Actual vs Predicted table on train data
mytable = mypd.crosstab(y_train,predy)
#Confusion matrix
mytable

col_0    0    1
Result         
0       92   13
1        3  134

**Analysing Confusion Matrix**

| | Predicted |  |
| ----------- | ----------- | ----------- |
| Actual|  0|1|
|0 |TN |FP |
|1 | FN |TP |


0 : Negative
1 : Positive

**Model accuracy from confusion matrix**

In [25]:
#True Positive
TP = 134
#False Positive
FP = 13
#True Negative
TN = 92
#False Negative
FN = 3

**Model accuracy**


In [26]:
#Accuracy
TP/(TP+FP)

0.9115646258503401

**Note:**

**Regression(Continuous Response Variable)**

Measure of model accuracy obtained from mymodel.score() is R<sup>2</sup> value

**Classification(Categorical Response Variable)**

Measure of model accuracy obtained from mymodel.score() is `accuracy` which is obtained from confusion matrix.

R2 is not a good measure to assess goodness of fit for a classification


In [27]:
#Misclassification rate
1 - TP/(TP+FP)

0.08843537414965985

Model accuracy on training data = 91.15% which is sufficiently good.

### Validate model on test data

In [28]:
#Get the predicted value on test data
predtest = mymodel.predict(X=x_test)

In [29]:
#Actual vs Predicted table on test data
mytable = mypd.crosstab(y_test,predtest)
#Confusion matrix
mytable

col_0    0   1
Result        
0       23  10
1        3  25

**Model accuracy from confusion matrix of test data**

In [30]:
#True Positive
TP = 25
#False Positive
FP = 10
#True Negative
TN = 23
#False Negative
FN = 3

**Model accuracy**


In [31]:
#Accuracy
TP/(TP+FP)

0.7142857142857143

In [32]:
#Misclassification rate
1 - TP/(TP+FP)

0.2857142857142857

Model accuracy on test data = 69.65% which is sufficient drop from 89.65% on train data.

| |Decision Tree Classifier Model(CART) | |
| ----------- | ----------- |----------- |
| | Train Data |Test Data |
|Accuracy |91.15 |71.42|
|Misclassification rate |8.84|28.57|


As there is a significant drop in model accuracy in test data, we can say that the model generalizability is not good.

## Bagging Method

Bagging generally uses all the p predictors.

`max_features='auto'`

In [33]:
#Specify a general bagging model
bag_model = RandomForestClassifier(n_estimators=500,min_samples_split=40,max_features='auto')

In [34]:
# Fit the bag model to the data
bag_model = bag_model.fit(x_train,y_train)

In [35]:
#Model accuarcy
accuracy = bag_model.score(x_train,y_train)
accuracy

0.8760330578512396

In [36]:
#Perform a 4 fold Cross Validation
myscore = cross_val_score(estimator=bag_model,X=x_train,y=y_train,cv=4,scoring='accuracy')

In [37]:
#Cross Validation Accuracy
myscore.mean()

0.8387978142076503

We see a considerably less drop in model accuracy after cross validation. We have to do hyper-parameter tuning for a more generalized model.

**Hyper parameter tuning**

In [38]:
#Hyper parameter tuning
myparamgrid = [{'min_samples_split':[5,10,15,20,25,30,35],'n_estimators':[100,200,300,400,500,600]}]

In [39]:
# Set up search
mysearch = GridSearchCV(estimator=bag_model,param_grid=myparamgrid,scoring='accuracy',cv=5,verbose=True)

In [40]:
#Fit the search on training data
mysearch = mysearch.fit(X=x_train,y=y_train)

Fitting 5 folds for each of 42 candidates, totalling 210 fits


In [40]:
#Optimum parameter values
mysearch.best_params_

{'min_samples_split': 10, 'n_estimators': 400}

In [41]:
#Specify the bagging model
bag_model = RandomForestClassifier(n_estimators=400,min_samples_split=10,max_features='auto')

In [42]:
# Fit the bag model to the data
bag_model = bag_model.fit(x_train,y_train)

In [43]:
#Model accuarcy
accuracy = bag_model.score(x_train,y_train)
accuracy

0.9504132231404959

In [44]:
#Perform a 4 fold Cross Validation
myscore = cross_val_score(estimator=bag_model,X=x_train,y=y_train,cv=4,scoring='accuracy')

In [45]:
#Cross Validation Accuracy
myscore.mean()

0.8262978142076502

In [46]:
# Predicted Values on train data
predy = bag_model.predict(X=x_train)

In [47]:
#Actual vs Predicted table on train data
mytable = mypd.crosstab(y_train,predy)
#Confusion matrix
mytable

col_0    0    1
Result         
0       96    9
1        3  134

**Model accuracy from confusion matrix of train data**

In [48]:
#True Positive
TP = 134
#False Positive
FP = 9
#True Negative
TN = 96
#False Negative
FN = 3

**Model accuracy**


In [49]:
#Accuracy
TP/(TP+FP)

0.9370629370629371

In [50]:
#Misclassification rate
1 - TP/(TP+FP)

0.06293706293706292

### Validate model on test data

In [51]:
#Get the predicted value on test data
predtest = bag_model.predict(X=x_test)

In [52]:
#Actual vs Predicted table on test data
mytable = mypd.crosstab(y_test,predtest)
#Confusion matrix
mytable

col_0    0   1
Result        
0       26   7
1        1  27

**Model accuracy from confusion matrix of test data**

In [53]:
#True Positive
TP = 27
#False Positive
FP = 7
#True Negative
TN = 26
#False Negative
FN = 1

**Model accuracy**


In [54]:
#Accuracy
TP/(TP+FP)

0.7941176470588235

In [55]:
#Misclassification rate
1 - TP/(TP+FP)

0.20588235294117652

Model accuracy on test data = 79.41% which is sufficient drop from 93.70% on train data.

| |Bagging Model | |
| ----------- | ----------- |----------- |
| | Train Data |Test Data |
|Accuracy |93.70 |79.41|
|Misclassification rate |6.29|20.59|


## Random Forest Method

Random Forest uses sqrt(p) predictors.

`max_features='sqrt'`

In [56]:
#Specify the bagging model
rf_model = RandomForestClassifier(n_estimators=500,min_samples_split=40,max_features='sqrt')

In [57]:
# Fit the bag model to the dat
rf_model= rf_model.fit(x_train,y_train)

In [58]:
#Model accuarcy
accuracy = rf_model.score(x_train,y_train)
accuracy

0.8677685950413223

In [59]:
#Perform a 4 fold Cross Validation
myscore = cross_val_score(estimator=rf_model,X=x_train,y=y_train,cv=4,scoring='accuracy')

In [60]:
#Cross Validation Accuracy
myscore.mean()

0.8346311475409837

We see a significant drop in model accuracy after cross validation. We have to do hyper-parameter tuning for a more generalized model.

**Hyper parameter tuning**

In [61]:
#Hyper parameter tuning
myparamgrid = [{'min_samples_split':[5,10,15,20,25,30,35],'n_estimators':[100,200,300,400,500,600]}]

In [62]:
# Set up search
mysearch = GridSearchCV(estimator=rf_model,param_grid=myparamgrid,scoring='accuracy',cv=5,verbose=True)

In [63]:
#Fit the search on training data
mysearch = mysearch.fit(X=x_train,y=y_train)

Fitting 5 folds for each of 42 candidates, totalling 210 fits


In [64]:
#Optimum parameter values
mysearch.best_params_

{'min_samples_split': 5, 'n_estimators': 100}

In [64]:
#Specify the bagging model
rf_model = RandomForestClassifier(n_estimators=100,min_samples_split=5,max_features='sqrt')

In [65]:
# Fit the bag model to the data
rf_model = rf_model.fit(x_train,y_train)

In [66]:
#Model accuarcy
accuracy = rf_model.score(x_train,y_train)
accuracy

0.9793388429752066

In [67]:
#Perform a 4 fold Cross Validation
myscore = cross_val_score(estimator=rf_model,X=x_train,y=y_train,cv=4,scoring='accuracy')

In [68]:
#Cross Validation Accuracy
myscore.mean()

0.8181010928961748

In [69]:
# Predicted Values on train data
predy = rf_model.predict(X=x_train)

In [70]:
#Actual vs Predicted table on train data
mytable = mypd.crosstab(y_train,predy)
#Confusion matrix
mytable

col_0     0    1
Result          
0       102    3
1         2  135

**Model accuracy from confusion matrix of train data**

In [71]:
#True Positive
TP = 135
#False Positive
FP = 3
#True Negative
TN = 102
#False Negative
FN = 2

**Model accuracy**


In [72]:
#Accuracy
TP/(TP+FP)

0.9782608695652174

In [73]:
#Misclassification rate
1 - TP/(TP+FP)

0.021739130434782594

### Validate model on test data

In [74]:
#Get the predicted value on test data
predtest = rf_model.predict(X=x_test)

In [75]:
#Actual vs Predicted table on test data
mytable = mypd.crosstab(y_test,predtest)
#Confusion matrix
mytable

col_0    0   1
Result        
0       26   7
1        3  25

**Model accuracy from confusion matrix of test data**

In [76]:
#True Positive
TP = 25
#False Positive
FP = 7
#True Negative
TN = 26
#False Negative
FN = 3

**Model accuracy**


In [77]:
#Accuracy
TP/(TP+FP)

0.78125

In [78]:
#Misclassification rate
1 - TP/(TP+FP)

0.21875

Model accuracy on test data = 78.12% which is sufficient drop from 97.82% on train data.

| |Random Forest Model | |
| ----------- | :-----------: |:-----------: |
| | Train Data |Test Data |
|Accuracy |97.82 |78.12|
|Misclassification rate |2.17|21.87|


**We comapare all three models**

| |Bagging Model |  |Random Forest| |CART| |
| ----------- | ----------- |----------- | ----------- |----------- |----------- |----------- |
| | Train Data |Test Data |Train Data |Test Data |Train Data |Test Data |
| Accuracy |93.70|79.41|97.82|78.12|91.15|71.42|
|Misclassification Rate |6.29 |20.59|2.17 |21.87|8.84|28.57|



When we compare the three model, CART has the worst performance on test data.

## Logistic Regression model

We will build logistic regression model on the dataset. We create train and test dataframes.


In [79]:
#We put another column in x_train & x_test: Intercept
x_train.loc[:,'Intercept'] = 1
x_test.loc[:,'Intercept'] = 1

/Users/sudiptarakshit/opt/anaconda3/lib/python3.8/site-packages/pandas/core/indexing.py:1597: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self.obj[key] = value
/Users/sudiptarakshit/opt/anaconda3/lib/python3.8/site-packages/pandas/core/indexing.py:1720: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_column(loc, value, pi)


In [80]:
#Check the train data
x_train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 242 entries, 184 to 8
Data columns (total 14 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Age          242 non-null    int64  
 1   Sex          242 non-null    int64  
 2   CP           242 non-null    int64  
 3   RestBP       242 non-null    int64  
 4   Cholesteral  242 non-null    int64  
 5   FBP          242 non-null    int64  
 6   RestECG      242 non-null    int64  
 7   Max_HR       242 non-null    int64  
 8   ExAngina     242 non-null    int64  
 9   Oldpeak      242 non-null    float64
 10  Slope        242 non-null    int64  
 11  CA           242 non-null    int64  
 12  Thal         242 non-null    int64  
 13  Intercept    242 non-null    int64  
dtypes: float64(1), int64(13)
memory usage: 28.4 KB


In [81]:
#Check the test data
x_test.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 61 entries, 69 to 252
Data columns (total 14 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Age          61 non-null     int64  
 1   Sex          61 non-null     int64  
 2   CP           61 non-null     int64  
 3   RestBP       61 non-null     int64  
 4   Cholesteral  61 non-null     int64  
 5   FBP          61 non-null     int64  
 6   RestECG      61 non-null     int64  
 7   Max_HR       61 non-null     int64  
 8   ExAngina     61 non-null     int64  
 9   Oldpeak      61 non-null     float64
 10  Slope        61 non-null     int64  
 11  CA           61 non-null     int64  
 12  Thal         61 non-null     int64  
 13  Intercept    61 non-null     int64  
dtypes: float64(1), int64(13)
memory usage: 7.1 KB


**Develop the Initial Model**

In [82]:
#Develop binary logistic regression model
mymodel = mysm.Logit(endog=y_train,exog=x_train).fit()

Optimization terminated successfully.
         Current function value: 0.364755
         Iterations 7


In [83]:
#Summary output
mymodel.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                 Result   No. Observations:                  242
Model:                          Logit   Df Residuals:                      228
Method:                           MLE   Df Model:                           13
Date:                Thu, 05 May 2022   Pseudo R-squ.:                  0.4670
Time:                        13:43:37   Log-Likelihood:                -88.271
converged:                       True   LL-Null:                       -165.62
Covariance Type:            nonrobust   LLR p-value:                 2.326e-26
===============================================================================
                  coef    std err          z      P>|z|      [0.025      0.975]
-------------------------------------------------------------------------------
Age             0.0046      0.026      0.176      0.860      -0.046       0.055
Sex            -1.7526      0.515     -3.400      0.001      -2.763      -0.742
CP              0.7819      0.201      3.888      0.000       0.388       1.176
RestBP         -0.0136      0.011     -1.195      0.232      -0.036       0.009
Cholesteral    -0.0104      0.005     -2.234      0.025      -0.019      -0.001
FBP            -0.3673      0.578     -0.635      0.525      -1.501       0.766
RestECG         0.1129      0.396      0.285      0.776      -0.663       0.889
Max_HR          0.0210      0.011      1.910      0.056      -0.001       0.043
ExAngina       -0.7361      0.446     -1.651      0.099      -1.610       0.138
Oldpeak        -0.5136      0.231     -2.221      0.026      -0.967      -0.060
Slope           0.6506      0.393      1.654      0.098      -0.120       1.421
CA             -0.7358      0.211     -3.493      0.000      -1.149      -0.323
Thal           -1.1005      0.330     -3.337      0.001      -1.747      -0.454
Intercept       4.5444      2.772      1.640      0.101      -0.888       9.977
===============================================================================
"""

In the model we see that the variables: Age, RestBP, Slope,ExAngina, FBP, RestECG are not significant features.

Only Sex, CP,CA,Thal,Max_HR,Oldpeak,Cholesteral are significant.

We develop new model using only these variables:Sex, CP,CA,Thal,Max_HR,Oldpeak,Cholesteral

In [84]:
#Overwrite x's in x train
x_train= x_train.iloc[:,[1,2,4,7,9,11,12,13]]

In [85]:
#Check train_data dataframe
x_train.head()

Sex  CP  Cholesteral  Max_HR  Oldpeak  CA  Thal  Intercept
184    1   0          243     128      2.6   0     3          1
19     0   3          239     151      1.8   2     2          1
118    0   1          204     172      0.0   0     2          1
41     1   1          245     180      0.2   0     2          1
59     0   0          303     159      0.0   1     2          1

In [86]:
#Develop the new binary logistic model
mymodel = mysm.Logit(endog=y_train,exog=x_train).fit()

Optimization terminated successfully.
         Current function value: 0.382225
         Iterations 7


In [87]:
#Check the model
mymodel.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                 Result   No. Observations:                  242
Model:                          Logit   Df Residuals:                      234
Method:                           MLE   Df Model:                            7
Date:                Thu, 05 May 2022   Pseudo R-squ.:                  0.4415
Time:                        13:44:01   Log-Likelihood:                -92.498
converged:                       True   LL-Null:                       -165.62
Covariance Type:            nonrobust   LLR p-value:                 2.496e-28
===============================================================================
                  coef    std err          z      P>|z|      [0.025      0.975]
-------------------------------------------------------------------------------
Sex            -1.6870      0.473     -3.565      0.000      -2.615      -0.759
CP              0.7961      0.184      4.324      0.000       0.435       1.157
Cholesteral    -0.0103      0.004     -2.443      0.015      -0.019      -0.002
Max_HR          0.0260      0.010      2.690      0.007       0.007       0.045
Oldpeak        -0.7548      0.200     -3.775      0.000      -1.147      -0.363
CA             -0.6969      0.198     -3.515      0.000      -1.085      -0.308
Thal           -1.0951      0.316     -3.462      0.001      -1.715      -0.475
Intercept       3.0918      1.846      1.674      0.094      -0.527       6.711
===============================================================================
"""

**Expression of the Model**

z = exp(y)/1+exp(y)

where 

y = 3.0918 -1.687*Sex + 0.796*CP - 0.697*CA - 1.095*Thal + 0.026*Max_HR - 0.0103*Cholesteral - 0.755*Oldpeak

**Model accuracy**

Here pseudo R2 value is 0.445. So, the model is not reasonably accurate.

**Confusion Matrix**

In [88]:
#Predicted Value
predy = mymodel.predict()
mynp.round(predy,3)

array([0.046, 0.881, 0.983, 0.882, 0.771, 0.081, 0.265, 0.31 , 0.811,
       0.016, 0.765, 0.671, 0.933, 0.954, 0.034, 0.887, 0.945, 0.825,
       0.158, 0.791, 0.409, 0.011, 0.992, 0.089, 0.817, 0.396, 0.058,
       0.861, 0.829, 0.12 , 0.832, 0.958, 0.814, 0.724, 0.453, 0.68 ,
       0.901, 0.503, 0.411, 0.983, 0.001, 0.062, 0.099, 0.141, 0.997,
       0.563, 0.663, 0.947, 0.599, 0.446, 0.124, 0.686, 0.761, 0.847,
       0.987, 0.093, 0.885, 0.542, 0.022, 0.206, 0.937, 0.04 , 0.402,
       0.959, 0.148, 0.996, 0.592, 0.719, 0.004, 0.678, 0.819, 0.748,
       0.006, 0.158, 0.958, 0.024, 0.928, 0.75 , 0.785, 0.778, 0.009,
       0.315, 0.936, 0.207, 0.015, 0.988, 0.769, 0.981, 0.533, 0.402,
       0.267, 0.065, 0.726, 0.884, 0.946, 0.02 , 0.991, 0.917, 0.772,
       0.88 , 0.084, 0.712, 0.791, 0.827, 0.934, 0.417, 0.796, 0.118,
       0.968, 0.793, 0.903, 0.529, 0.011, 0.01 , 0.95 , 0.956, 0.063,
       0.957, 0.034, 0.381, 0.967, 0.918, 0.766, 0.779, 0.97 , 0.978,
       0.129, 0.682,

In [89]:
#Predicted class
predclass = [1 if i>0.5 else 0 for i in predy]

In [90]:
#Create dataframe for predclass
predclass = mypd.DataFrame(predclass,columns=['Predicted_Class'])

In [91]:
#Create the confusion matrix
mytable = confusion_matrix(y_train,predclass.Predicted_Class)
mytable

array([[ 82,  23],
       [ 15, 122]])

In [92]:
#True Positive
TP = 122
#False Positive
FP = 23
#True Negative
TN = 82
#False Negative
FN = 15

In [93]:
#Accuarcy
Accuracy = (TP + TN)/(TP+TN+FP+FN)
round(Accuracy,3)

0.843

### Validate model on test data

In [94]:
#Overwrite x's in x_test
x_test= x_test.iloc[:,[1,2,4,7,9,11,12,13]]

In [95]:
#Get the predicted value on test data
predtest = mymodel.predict(x_test)

In [96]:
#Predicted class
predtestclass = [1 if i>0.5 else 0 for i in predtest]

In [97]:
#Create the confusion matrix
mytable = confusion_matrix(y_test,predtestclass)
mytable

array([[26,  7],
       [ 3, 25]])

In [98]:
#True Positive
TP = 25
#False Positive
FP = 7
#True Negative
TN = 26
#False Negative
FN = 3

In [99]:
#Accuarcy
Accuracy = (TP + TN)/(TP+TN+FP+FN)
round(Accuracy,3)

0.836

| |Logistic Regression Model | |
| ----------- | ----------- |----------- |
| | Train Data |Test Data |
|Accuracy |84.3 |83.6|


We see model generalizibility is very good in case of logistic regression.